include/rmm/mr/device/cuda_async_memory_resource.hpp

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <rmm/detail/error.hpp>
+#include <rmm/mr/device/device_memory_resource.hpp>
+#include "rmm/cuda_stream_view.hpp"
+
+#include <cuda_runtime_api.h>
+
+namespace rmm {
+namespace mr {
+
+#if CUDART_VERSION >= 11020  // 11.2 introduced cudaMallocAsync
+
+/**
+ * @brief `device_memory_resource` derived class that uses `cudaMallocAsync`/`cudaFreeAsync` for
+ * allocation/deallocation.
+ */
+class cuda_async_memory_resource final : public device_memory_resource {
+ public:
+  /**
+   * @brief Default constructor
+   *
+   * @throws rmm::runtime_error if the CUDA version does not support `cudaMallocAsync`
+   */
+  cuda_async_memory_resource()
+  {
+    // Check if cudaMallocAsync Memory pool supported
+    int device{0};
+    RMM_CUDA_TRY(cudaGetDevice(&device));
+    int v;
+    RMM_CUDA_TRY(cudaDeviceGetAttribute(&v, cudaDevAttrMemoryPoolsSupported, device));
```

Comment:
If someone compiles with CUDA 11.2 and then runs on CUDA 11.0, wouldn't we fail to find the `cudaDevAttrMemoryPoolsSupported` symbol from the lack of runtime version support?

Reply:
`cudaDevAttrMemoryPoolsSupported` exists at compile time. In the binary, only its value will be stored. At runtime, this value will be passed to `cudaDeviceGetAttribute`, which will then return `cudaErrorInvalidValue`, which we take to mean that memory pools are not supported and we throw an error.

I could clean this up to throw the same error no matter what CUDA returns so I don't have two possible error messages, though.

Reply:
Unified the exception here.